In [ ]:
Dining philosopher's problem:
5 forks , 5 philosophers.
only 2 can be server at a time. all 5 cannont be served simultaneously.
if the forks are shared quickly enough with least possible time, it might seem noone is left out..

In [ ]:
Thread is a line of code execution. 
Each core can run 1 thread at a time.



A core will run a thread and the wrapper or the package is a process.
A process contains atleast one thread + some resources set aside for operating system(like a networks connection, hard drive
                                                                                     file pointers etc.)

These resources continuously change. 
Thread is what runs and process is what manages all the resources that is needed to run.


In [ ]:
Python GIL 
        GLOBAL INTERPRETER LOCK
    The Python Global Interpreter Lock or GIL, in simple words, is a mutex (or a lock) that allows only one thread to hold the
    control of the Python interpreter.
    This means that only one thread can be in a state of execution at any point in time.
    
    
What is the benefit of making more threads if only one thread can run in a core at a time?
Because there can be more cores.

Each process creates a key resources or critical resource. It must acquire it. This key resources is called GIL.

What is the point of threads?
Lets say u have 2 programs.
one does complex calculation and takes long time.
ANother does user interface functions. This can take long time since user can take long time to type/input.

1. Computer is being used  for a long time.
2. Computer is waiting for a long time.

Solution : CO-OPERATIVE MULTITASKING.
    Ask the user and wait for input and immediately releases GIL and other can acquire it and run and then release for first 
    program to use and so on.
    
    
THREAD IN PYTHON IS TO REDUCE WAITING TIME.


In [2]:
import time


def ask_user():
    s1=time.time()
    user_name=input('Enter your name: ')
    greet=f'Hello {user_name} , Welcome to XYZ'
    print(greet)
    print(f'ask_user time is {time.time()-start}')
    
def calc():
    s2=time.time()
    print('Calculating...')
    b=[x**2 for x in range(1000000)]
    print(f'calc time is {time.time()-s2}')
    
start=time.time()    
ask_user()
calc()
print(f'total time is {time.time()-start}')

Enter your name: prashanth
Hello prashanth , Welcome to XYZ
ask_user time is 2.929628610610962
Calculating...
calc time is 0.5349011421203613
total time is 3.479698419570923


In [3]:
import time


def ask_user():
    s1=time.time()
    user_name=input('Enter your name: ')
    greet=f'Hello {user_name} , Welcome to XYZ'
    print(greet)
    print(f'ask_user time is {time.time()-start}')
    
def calc():
    s2=time.time()
    print('Calculating...')
    b=[x**2 for x in range(1000000)]
    print(f'calc time is {time.time()-s2}')
    
start=time.time()    
ask_user()
calc()
print(f'total time is {time.time()-start}')

Enter your name: prashanth
Hello prashanth , Welcome to XYZ
ask_user time is 12.692781209945679
Calculating...
calc time is 0.5273876190185547
total time is 13.239768266677856


In [ ]:
About 2 cases were run SEQUENTIALLY.

In [5]:
import time
from threading import Thread


def ask_user():
    s1=time.time()
    user_name=input('Enter your name: ')
    greet=f'Hello {user_name} '
    print(greet)
    print(f'ask_user time is {time.time()-start}')
    
def calc():
    s2=time.time()
    print('Calculating...')
    b=[x**2 for x in range(1000000)]
    print(f'calc time is {time.time()-s2}')
    
start=time.time()    
ask_user()
calc()
print(f'total time is {time.time()-start}')



thread1=Thread(target=ask_user)
thread2=Thread(target=calc)


t3=time.time()
thread1.start()
thread2.start()
# 3 threads.. 1 main and these 2.. 
thread1.join()
thread2.join()
#blocking operations.. THis will make main thread to wait these 2 are finished.

print(f'2 threads total time {time.time()-start}')

Enter your name: prashanth
Hello prashanth , Welcome to XYZ
ask_user time is 3.623936653137207
Calculating...
calc time is 0.4890298843383789
total time is 4.1598498821258545
Calculating...
calc time is 0.701505184173584
Enter your name: prashanth
Hello prashanth , Welcome to XYZ
ask_user time is 8.233621835708618
2 threads total time 8.233621835708618


In [6]:
import time
from threading import Thread


def ask_user():
    s1=time.time()
    user_name=input('Enter your name: ')
    greet=f'Hello {user_name} '
    print(greet)
    print(f'ask_user time is {time.time()-start}')
    
def calc():
    s2=time.time()
    print('Calculating...')
    b=[x**2 for x in range(1000000)]
    print(f'calc time is {time.time()-s2}')
    
start=time.time()    
ask_user()
calc()
print(f'total time is {time.time()-start}')



thread1=Thread(target=calc)
thread2=Thread(target=calc)


t3=time.time()
thread1.start()
thread2.start()
# 3 threads.. 1 main and these 2.. 
thread1.join()
thread2.join()
#blocking operations.. THis will make main thread to wait these 2 are finished.

print(f'2 threads total time {time.time()-start}')

Enter your name: prasha
Hello prasha 
ask_user time is 2.181922197341919
Calculating...
calc time is 0.49190640449523926
total time is 2.689457416534424
Calculating...
Calculating...
calc time is 0.8489410877227783
calc time is 0.9832792282104492
2 threads total time 3.7508602142333984


note , threading for complex calculation has actually doubled the time ( 0.49 to 0.84 and 0.983) This is because there is time
being used up removing and adding thread alternatively. Hence dont use threads if a process requires CPU all the time.

In [ ]:
from concurrent.futures import ThreadPoolExecutor   #famous module....

with ThreadPoolExecutor(max_workers=2) as pool:  #creates a pool of empty threads and can be used later dynamically
    pool.submit(calc)
    pool.submit(ask_user)

In [ ]:
KILLING THE THREADS :
    Manually killing the threads while they are still executing can kill the resources (GIL) if they are not
    released before . Other threads cant pick up the GIL that was assigned to the first one.  #DEADLOCK
    

In [ ]:
MULTIPROCESSING:
        

In [9]:
from multiprocessing import Process
process1=Process(target=calc)
process1.start()
s1=time.time()
ask_user()
process1.join()
print(f'2 process time is {time.time()-s1}')


Enter your name: prashanth
Hello prashanth 
ask_user time is 973.509304523468
2 process time is 2.496448278427124


In [2]:
from multiprocessing import Process
import time


def ask_user():
    s1=time.time()
    user_name=input('Enter your name: ')
    greet=f'Hello {user_name} , Welcome to XYZ'
    print(greet)
    print(f'ask_user time is {time.time()-start}')
    
def calc():
    s2=time.time()
    print('Calculating...')
    b=[x**10 for x in range(1000000)]
    print(f'calc time is {time.time()-s2}')

process1=Process(target=calc)
process2=Process(target=ask_user)
process1.start()
process2.start()

s1=time.time()

process1.join()
process2.join()

print(f'2 process time is {time.time()-s1}')


2 process time is 0.24804449081420898


In [2]:
import time
from threading import Thread

counter=0

def counterplus():
    global counter
    counter+=1
    print(f'New Counter Value : {counter}')
    print('------------------')
          
for i in range(10):
    counterplus()
    

New Counter Value : 1
------------------
New Counter Value : 2
------------------
New Counter Value : 3
------------------
New Counter Value : 4
------------------
New Counter Value : 5
------------------
New Counter Value : 6
------------------
New Counter Value : 7
------------------
New Counter Value : 8
------------------
New Counter Value : 9
------------------
New Counter Value : 10
------------------


In [3]:
import time
from threading import Thread

counter=0

def counterplus():
    global counter
    counter+=1
    print(f'New Counter Value : {counter}')
    print('------------------')
          
for i in range(10):
    t=Thread(target=counterplus)
    t.start()
    

New Counter Value : 1
------------------
New Counter Value : 2
------------------
New Counter Value : 3
------------------
New Counter Value : 4
------------------New Counter Value : 5
------------------
New Counter Value : 6

------------------
New Counter Value : 7
------------------
New Counter Value : 8
------------------
New Counter Value : 9
New Counter Value : 10
------------------
------------------


In [ ]:
Solution Thread+queue module = sequetial execution

In [ ]:
A Class in Python has three types of access modifiers – 

Public Access Modifier
Protected Access Modifier
Private Access Modifier

Public Access Modifier:
The members of a class that are declared public are easily accessible from any part of the program. All data members
and member functions of a class are public by default.



Python uses ‘_’ symbol to determine the access control for a specific data member or a member function of a class. Access 
specifiers in Python have an important role to play in securing data from unauthorized access and in 
preventing it from being exploited.


Public Access Modifier:
The members of a class that are declared public are easily accessible from any part of the program. 
All data members and member functions of a class are public by default.

Protected Access Modifier:
The members of a class that are declared protected are only accessible to a class derived from it. 
Data members of a class are declared protected by adding a single underscore ‘_’ symbol before the data member of that class.

Private Access Modifier:
The members of a class that are declared private are accessible within the class only, 
K1
private access modifier is the most secure access modifier. 
Data members of a class are declared private by adding a double underscore ‘__’ symbol before the data member of that class.


In [10]:
#ACCESS MODIFIERS 


class Super: 
    var1 = None
    _var2 = None
    __var3 = None
      
    
    def __init__(self, var1, var2, var3):   
        self.var1 = var1 
        self._var2 = var2 
        self.__var3 = var3 
      
      
    def displayPublicMembers(self): 
   
          
        print("Public Data Member: ", self.var1) 
         
        
    def _displayProtectedMembers(self): 
   
           
        print("Protected Data Member: ", self._var2) 
       
    
    def __displayPrivateMembers(self): 
        print("Private Data Member: ", self.__var3) 
  
     
    def __accessPrivateMembers(self):
        self.__displayPrivateMembers() 
   
 
class Sub(Super): 
    def __init__(self, var1, var2, var3):
        Super.__init__(self, var1, var2, var3) 
        
    def test(self):
        print(f'{self.var1}')
        print(f'{self._var2}')
        print(f'{self.__var3}')
            
       
    def accessProtectedMemebers(self):
        self._displayProtectedMembers() 
   
     
obj = Sub("Geeks", 4, "Geeks !") 

obj.test()
  
obj.displayPublicMembers() 
obj.accessProtectedMemebers() 
obj.accessPrivateMembers()  
  

print("Object is accessing protected member:", obj._var2) 
  






Geeks
4


AttributeError: 'Sub' object has no attribute '_Sub__var3'

In [14]:
#USING YIELD TO RECEIVE DATA AS OPPOSED TO ONLY GIVE IT BACK


def greet():
    name=yield
    print(f'Hello {name}')
    
g=greet()
g.send(None) #Priming the generator . This runs the greet function right upto yield and suspend the function.
g.send('Adam')



Hello Adam


StopIteration: 

In [22]:
from collections import deque

friends=deque(('Ras','Mael','Estarossa'))

def friend_upper():
    while friends:
        friend=friends.popleft().upper()
        greeting=yield    
        #yield from greeting
        print(f'{greeting} {friend}')
        
def greet(g):
    g.send(None)
    while True:
        greeting=yield
        #yield from greeting
        g.send(greeting)
        
greeter=greet(friend_upper())
greeter.send(None)
greeter.send("Hello")

print('Next command.....')

greeter.send('How are You? ')



Hello RAS
Next command.....
How are You?  MAEL


These are called Co-routines.
No more a generator..

In [24]:
from collections import deque
from types import coroutine

friends=deque(('Ras','Mael','Estarossa'))



@coroutine
def friend_upper():
    while friends:
        friend=friends.popleft().upper()
        greeting=yield    
        #yield from greeting
        print(f'{greeting} {friend}')
        
async def greet(g):
    
    await g
   
        
greeter=greet(friend_upper())
greeter.send(None)
greeter.send("Hello")

print('Next command.....')

greeter.send('How are You? ')



Hello RAS
Next command.....
How are You?  MAEL


In [26]:
from collections import deque
from types import coroutine

friends=deque(('Ras','Mael','Estarossa'))



@coroutine
def friend_upper():
    while friends:
        friend=friends.popleft().upper()
        greeting=yield    
        #yield from greeting
        print(f'{greeting} {friend}')
        
async def greet(g):
    print('Starting....')
    await g
    print('Ending....')
   
        
greeter=greet(friend_upper())
greeter.send(None)
greeter.send("Hello")

print('Next command.....')

greeter.send('How are You? ')
greeter.send('How are You? ')
greeter.send('How are You? ')




Starting....
Hello RAS
Next command.....
How are You?  MAEL
How are You?  ESTAROSSA
Ending....


StopIteration: 

In [ ]:
Explanation....greeter.hello starts greet and suspends it and initiating friend_upper. Hence, ending is not printed cos
the coroutine hasnt finished. It will finish only when the coroutine has gone through the list its iterating over.


In [30]:
from collections import deque
from types import coroutine

friends=deque(('Ras','Mael','Estarossa'))



@coroutine
def friend_upper():
    while friends:
        friend=friends.popleft().upper()
        greeting=yield    
        #yield from greeting
        print(f'{greeting} {friend}')
        
async def greet(g):
    print('Starting....')
    await g
    print('Ending....')
   
        
greeter=greet(friend_upper())
greeter.send(None)
greeting=input('Enter a greeting : ')
greeter.send(greeting)
greeting=input('Enter a greeting : ')
greeter.send(greeting)




Starting....
Enter a greeting : Hello
Hello RAS
Enter a greeting : how are you
how are you MAEL


In [36]:
#titbit
num='seku'
num.ljust(10,'0')   #num.rjust(1)


'seku000000'